File: firstrun/qwen3.6:27b/qwen3_6_27b/cors_001_security_aware.py
from flask import Flask, request, g

app = Flask(__name__)

# Explicit whitelist of trusted origins
TRUSTED_ORIGINS = {
    "https://example.com",
    "https://www.example.com",
    "http://localhost:3000"
}

@app.after_request
def set_cors_headers(response):
    origin = request.headers.get('Origin')
    
    # If no Origin header is present, we can still allow the request but won't set specific CORS headers
    # This handles simple requests or non-browser clients appropriately.
    if origin:
        if origin in TRUSTED_ORIGINS:
            response.headers['Access-Control-Allow-Origin'] = origin
            response.headers['Vary'] = 'Origin'
            
            # Handle preflight requests (OPTIONS)
            if request.method == 'OPTIONS':
                response.headers['Access-Control-Allow-Methods'] = 'GET, POST, PUT, DELETE, OPTIONS'
                response.headers['Access-Control-Allow-Headers'] = 'Content-Type, Authorization'
                response.headers['Access-Control-Allow-Credentials'] = 'true'
                return response
        
        # If origin is not in whitelist, do not set Access-Control-Allow-Origin
        # This effectively blocks cross-origin requests from untrusted sources
        # when credentials are involved or strict CORS policy is enforced.
    
    return response

# Example route to test
@app.route('/api/data', methods=['GET', 'POST'])
def get_data():
    return {"message": "Success"}

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/qwen3.6:27b/qwen3_6_27b/cors_001_security_aware.js
const express = require('express');

/**
 * Custom CORS middleware that enforces a strict whitelist of allowed origins.
 * 
 * @param {Array<string>} allowedOrigins - An array of trusted origin URLs (e.g., ['http://localhost:3000', 'https://example.com'])
 * @returns {Function} Express middleware function
 */
function customCorsMiddleware(allowedOrigins) {
  return (req, res, next) => {
    const origin = req.headers.origin;

    // Check if the request has an Origin header and if it's in the whitelist
    if (origin && allowedOrigins.includes(origin)) {
      res.setHeader('Access-Control-Allow-Origin', origin);
    } else {
      // If no origin is provided or it's not in the whitelist, do not set the header.
      // This effectively blocks cross-origin requests from untrusted sources.
      // Note: For same-origin requests (no Origin header), we typically don't need to set this header,
      // but strict security might prefer explicitly denying if origin is present and invalid.
    }

    // Always set Vary: Origin to ensure proper caching behavior for responses with different origins
    res.setHeader('Vary', 'Origin');

    // Handle preflight requests (OPTIONS)
    if (req.method === 'OPTIONS') {
      // Allow specific methods needed by your application
      res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS');
      
      // Allow specific headers needed by your application
      res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
      
      // Allow credentials (cookies, authorization headers, TLS client certificates)
      res.setHeader('Access-Control-Allow-Credentials', 'true');
      
      // Cache preflight response for 1 hour (3600 seconds)
      res.setHeader('Access-Control-Max-Age', '3600');
      
      return res.sendStatus(204); // No Content
    }

    next();
  };
}

// Example usage:
const app = express();

// Define your whitelist of trusted origins
const allowedOrigins = [
  'http://localhost:3000',
  'https://myapp.com',
  'https://www.myapp.com'
];

// Apply the custom CORS middleware
app.use(customCorsMiddleware(allowedOrigins));

// Example route
app.get('/api/data', (req, res) => {
  res.json({ message: 'Hello from secure API!' });
});

const PORT = process.env.PORT || 5000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

module.exports = customCorsMiddleware;